src/test/java/com/google/sps/servlets/SearchServletTest.java

```diff
@@ -0,0 +1,216 @@
+    expectedFilters.add(new FilterPredicate("professor", FilterOperator.EQUAL, "Smith"));
+    expectedFilters.add(new FilterPredicate("term", FilterOperator.EQUAL, "Spring 2020"));
+    expectedFilters.add(new FilterPredicate("units", FilterOperator.IN, units));
+
+    List<Filter> filters = searchObject.getFilters(request);
+
+    assertEquals(filters, expectedFilters);
+  }
+
+  @Test
+  /* Ensure no filters are made given no search parameters are specified. */
+  public void MakeFilters_NoParamsSet() {
+
+    // TODO: Find a better way to replicate null parameters -
+    // getting NullPointerException when trying other methods.
+    request = createRequest(request, "", "", "select", "");
+
+    List<Filter> expectedFilters = new ArrayList<>();
+
+    List<Filter> filters = searchObject.getFilters(request);
+
+    assertEquals(filters, expectedFilters);
+  }
+
+  @Test
+  /* Ensure the correct results are retrieved from Datastore given that all filters are set. */
+  public void GetResults_AllFiltersSet() {
+
+    request = createRequest(request, "CS 105", "Smith", "Spring 2020", "1,2,3");
+    List<Entity> expectedResults = new ArrayList<>();
+
+    addCourseEntityWithList("CS 105", "Smith", "Spring 2020", 1, expectedResults, true);
+    addCourseEntityWithList("CS 105", "Smith", "Spring 2020", 4, expectedResults, false);
+
+    List<Filter> filters = searchObject.getFilters(request); // Tested in makeFiltersAllParams.
+    List<Entity> results = searchObject.getResults(filters);
+
+    assertEquals(results, expectedResults);
+  }
+
+  @Test
+  /* Ensure the correct results are retrieved from Datastore given no search filters. */
+  public void GetResults_NoFiltersSet() {
+
+    request = createRequest(request, "", "", "select", "");
+    List<Entity> expectedResults = new ArrayList<>();
+
+    addCourseEntityWithList("CS 105", "Smith", "Spring 2020", 1, expectedResults, true);
+    addCourseEntityWithList("CS 105", "Smith", "Spring 2020", 4, expectedResults, true);
+
+    List<Filter> filters = searchObject.getFilters(request); // Tested in makeFiltersNoParams.
+    List<Entity> results = searchObject.getResults(filters);
+
+    assertEquals(results, expectedResults);
+  }
+
+  @Test
+  /* Ensure the whole doGet process works given all search filters are set. */
+  public void GetCourses_AllParamsSet() {
+
+    request = createRequest(request, "CS 105", "Smith", "Spring 2020", "1,2,3");
+    List<Course> expectedCourses = new ArrayList<>();
+
+    addCourseEntity("CS 105", "Smith", "Spring 2020", 1);
+    expectedCourses.add(new Course("CS 105", "Smith", Long.valueOf(1), "Spring 2020"));
+
+    addCourseEntity("CS 105", "Smith", "Spring 2020", 2);
+    expectedCourses.add(new Course("CS 105", "Smith", Long.valueOf(2), "Spring 2020"));
+
+    addCourseEntity("CS 106", "Smith", "Spring 2020", 3);
+
+    List<Filter> filters = searchObject.getFilters(request); // Tested in makeFiltersAllParams.
```

Comment:
Ideally this is tested just via `doGet` since that should likely be the input to the SearchServlet "black box".

Reply:
Makes sense, I just made a getHelper function that essentially does everything up to the converting into json and tested that - is that sufficient, or would we want to test the json conversion as well?